src/container/tenant_builder.rs

```diff
@@ -156,28 +159,72 @@ impl<'a> TenantContainerBuilder<'a> {
         Ok(container)
     }
 
-    fn adapt_spec_for_tenant(&self, spec: &mut Spec, container: &Container) -> Result<()> {
-        if let Some(ref process) = self.process {
-            self.set_process(spec, process)?;
+    fn adapt_spec_for_tenant(&self, spec: &Spec, container: &Container) -> Result<Spec> {
+        let process = if let Some(ref process) = self.process {
+            self.set_process(process)?
         } else {
-            self.set_working_dir(spec)?;
-            self.set_args(spec)?;
-            self.set_environment(spec)?;
-            self.set_no_new_privileges(spec)?;
-            self.set_capabilities(spec)?;
-        }
+            let mut process_builder = ProcessBuilder::default();
+
+            process_builder = match self.set_working_dir()? {
+                Some(cwd) => process_builder.cwd(cwd),
+                None => process_builder,
+            };
+
+            process_builder = process_builder.args(self.set_args()?);
+            process_builder = process_builder.env(self.set_environment()?);
+
+            process_builder = match self.set_no_new_privileges() {
+                Some(no_new_priv) => process_builder.no_new_privileges(no_new_priv),
+                None => process_builder,
+            };
+
+            process_builder = match self.set_capabilities(spec)? {
+                Some(caps) => process_builder.capabilities(caps),
+                None => process_builder,
+            };
+
+            process_builder.build()?
+        };
 
         if container.pid().is_none() {
             bail!("Could not retrieve container init pid");
         }
 
         let init_process = procfs::process::Process::new(container.pid().unwrap().as_raw())?;
-        self.set_namespaces(spec, init_process.namespaces()?)?;
+        let ns = self.set_namespaces(init_process.namespaces()?)?;
+        let linux = LinuxBuilder::default().namespaces(ns).build()?;
 
-        Ok(())
+        let mut spec_builder = SpecBuilder::default()
```

Comment:
This is fine, but I think we should implement in `oci-spec-rs` From spec to spec_builder. Otherwise, there will just be a lot of boilerplate code.

Reply:
I think so too.